File: mobile/apps/photos/plugins/ente_photos_platform/android/src/main/kotlin/io/ente/photos/platform/MlProcessLockRegistry.kt
package io.ente.photos.platform

internal data class MlProcessLockState(
    val token: String,
    val pluginInstanceId: String,
    val origin: String,
    val operation: String,
    val acquiredAtNanos: Long,
)

internal data class MlProcessLockAcquireResult(
    val acquired: Boolean,
    val holder: MlProcessLockState,
)

internal object MlProcessLockRegistry {
    private val monitor = Any()
    private var holder: MlProcessLockState? = null

    fun tryAcquire(
        pluginInstanceId: String,
        token: String,
        origin: String,
        operation: String,
        acquiredAtNanos: Long = System.nanoTime(),
    ): MlProcessLockAcquireResult =
        synchronized(monitor) {
            val current = holder
            if (current == null) {
                val acquired =
                    MlProcessLockState(
                        token = token,
                        pluginInstanceId = pluginInstanceId,
                        origin = origin,
                        operation = operation,
                        acquiredAtNanos = acquiredAtNanos,
                    )
                holder = acquired
                MlProcessLockAcquireResult(true, acquired)
            } else {
                MlProcessLockAcquireResult(
                    current.pluginInstanceId == pluginInstanceId && current.token == token,
                    current,
                )
            }
        }

    fun release(pluginInstanceId: String, token: String): Boolean =
        synchronized(monitor) {
            val current = holder
            if (current?.pluginInstanceId != pluginInstanceId || current.token != token) {
                return@synchronized false
            }
            holder = null
            true
        }

    fun resetForInstance(pluginInstanceId: String): Boolean =
        synchronized(monitor) {
            if (holder?.pluginInstanceId != pluginInstanceId) {
                return@synchronized false
            }
            holder = null
            true
        }

    fun state(): MlProcessLockState? = synchronized(monitor) { holder }
}

File: mobile/apps/photos/plugins/ente_photos_platform/ios/Sources/MlProcessLockCore/MlProcessLockRegistry.swift
import Foundation

struct MlProcessLockState: Equatable {
    let token: String
    let pluginInstanceID: String
    let origin: String
    let operation: String
    let acquiredAtUptimeNanoseconds: UInt64
}

struct MlProcessLockAcquireResult {
    let acquired: Bool
    let holder: MlProcessLockState
}

final class MlProcessLockRegistry: @unchecked Sendable {
    static let shared = MlProcessLockRegistry()

    private let lock = NSLock()
    private var holder: MlProcessLockState?

    func tryAcquire(
        pluginInstanceID: String,
        token: String,
        origin: String,
        operation: String,
        acquiredAtUptimeNanoseconds: UInt64 = DispatchTime.now().uptimeNanoseconds
    ) -> MlProcessLockAcquireResult {
        lock.lock()
        defer { lock.unlock() }

        if let holder {
            return MlProcessLockAcquireResult(
                acquired: holder.pluginInstanceID == pluginInstanceID && holder.token == token,
                holder: holder
            )
        }
        let acquired = MlProcessLockState(
            token: token,
            pluginInstanceID: pluginInstanceID,
            origin: origin,
            operation: operation,
            acquiredAtUptimeNanoseconds: acquiredAtUptimeNanoseconds
        )
        holder = acquired
        return MlProcessLockAcquireResult(acquired: true, holder: acquired)
    }

    func release(pluginInstanceID: String, token: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }

        guard holder?.pluginInstanceID == pluginInstanceID, holder?.token == token else {
            return false
        }
        holder = nil
        return true
    }

    func reset(for pluginInstanceID: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }

        guard holder?.pluginInstanceID == pluginInstanceID else {
            return false
        }
        holder = nil
        return true
    }

    func state() -> MlProcessLockState? {
        lock.lock()
        defer { lock.unlock() }
        return holder
    }
}
